dm/worker/task_checker.go

```diff
@@ -0,0 +1,296 @@
+	defer ticker.Stop()
+	for {
+		select {
+		case <-tsc.ctx.Done():
+			tsc.l.Info("worker task checker exits")
+			return
+		case <-ticker.C:
+			tsc.check()
+		}
+	}
+}
+
+// isResumableError checks the error message and returns whether we need to
+// resume the task and retry
+func isResumableError(err *pb.ProcessError) bool {
+	// not elegant code, because TiDB doesn't expose some error
+	unsupportedDDLMsgs := []string{
+		"can't drop column with index",
+		"unsupported add column",
+		"unsupported modify column",
+		"unsupported modify",
+		"unsupported drop integer primary key",
+	}
+	if err.Type == pb.ErrorType_ExecSQL {
+		for _, msg := range unsupportedDDLMsgs {
+			if strings.Contains(err.Msg, msg) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (tsc *realTaskStatusChecker) getResumeStrategy(stStatus *pb.SubTaskStatus, duration time.Duration) ResumeStrategy {
+	// task that is not paused or paused manually, just ignore it
+	if stStatus == nil || stStatus.Stage != pb.Stage_Paused || stStatus.Result == nil || stStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range stStatus.Result.Errors {
+		if !isResumableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.latestResumeTime[stStatus.Name]) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	allSubTaskStatus := tsc.w.getAllSubTaskStatus()
+	tasks := make(map[string]struct{}, len(allSubTaskStatus))
+
+	defer func() {
+		// cleanup outdated tasks
+		for taskName := range tsc.backoffs {
+			_, ok := tasks[taskName]
+			if !ok {
+				tsc.l.Debug("remove task from checker", zap.String("task", taskName))
+				delete(tsc.backoffs, taskName)
+				delete(tsc.latestPausedTime, taskName)
+				delete(tsc.latestBlockTime, taskName)
+				delete(tsc.latestResumeTime, taskName)
+			}
+		}
+	}()
+
+	for _, stStatus := range allSubTaskStatus {
+		taskName := stStatus.Name
+		tasks[taskName] = struct{}{}
+		bf, ok := tsc.backoffs[taskName]
+		if !ok {
+			bf, _ = backoff.NewBackoff(tsc.cfg.BackoffFactor, tsc.cfg.BackoffJitter, tsc.cfg.BackoffMin, tsc.cfg.BackoffMax)
+			tsc.backoffs[taskName] = bf
+			tsc.latestPausedTime[taskName] = time.Now()
+			tsc.latestResumeTime[taskName] = time.Now()
+		}
+		duration := bf.Current()
+		strategy := tsc.getResumeStrategy(stStatus, duration)
+		switch strategy {
+		case ResumeIgnore:
+			if time.Since(tsc.latestPausedTime[taskName]) > tsc.cfg.BackoffRollback {
+				bf.Rollback()
+				// after each rollback, reset this timer
+				tsc.latestPausedTime[taskName] = time.Now()
+			}
+		case ResumeNoSense:
+			// this strategy doesn't forward or rollback backoff
+			tsc.latestPausedTime[taskName] = time.Now()
+			blockTime, ok := tsc.latestBlockTime[taskName]
```

Comment:
`latestBlockTime` only used for log? can we remove it

Reply:
if removed, we can't know how long this task blocked